File: frame/map-frame-project/MapFrameCoreModule/src/main/java/com/grandtech/mapframe/core/sketch/SketchHelpView.java
package com.grandtech.mapframe.core.sketch;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PointF;

import android.graphics.Rect;
import android.util.AttributeSet;
import android.widget.FrameLayout;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.grandtech.mapframe.core.R;
import com.grandtech.mapframe.core.util.MeasureUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created by zy on 2018/5/29.
 */

public final class SketchHelpView extends FrameLayout {

    /**
     * 绘制拖拽时候的线和面
     */
    public static final int DRAG_POLYGON_POLYLINE = 1000;
    /**
     * 绘制十字丝
     */
    public static final int DRAW_TARGET = 2000;
    /**
     * 绘制帮助点
     */
    public static final int DRAG_POINT = 3000;
    /**
     * 绘制橡皮绳 释放
     */
    public static final int DRAG_NORMAL = 4000;

    /**
     *
     */
    public static final int DRAG_NOTHING = 5000;
    private int _flag;
    private Context _context;

    private String[] _text;
    /**
     * 屏幕中心坐标
     */
    private PointF _screenCenterPoint;
    /**
     *
     */
    private Paint _paint;
    /**
     *
     */
    private int _alpha = 255;
    /**
     * 圆圈半径
     */
    private int _radius = 5;

    private int _textSize = 30;
    /**
     * 绘制图形
     */
    private Path _path;
    //*************************橡皮绳*************************
    /**
     * 坐标一
     */
    private PointF _screenPoint1;
    /**
     * 坐标二
     */
    private PointF _screenPoint2;
    //***********************拖拽和平移***********************

    private Bitmap _targetBitmap;
    private Bitmap _crossHair;
    /**
     * 平移矢量
     */
    private float _vx = 0;
    private float _vy = 0;
    private PointF _bPoint;
    private PointF _aPoint;
    private PointF _cPoint;
    //拖拽点的完成结果
    private PointF _mbPoint;
    /**
     * 图形点集合
     */
    private List<PointF> _oPointFS;
    /**
     * 平移后的
     */
    private List<PointF> _mPointFS;
    private boolean _isClose = true;
    private boolean _isRelease;

    //***********************拖拽和平移***********************

    public SketchHelpView(@NonNull Context context) {
        super(context);
        this._context = context;
        init();
    }

    public SketchHelpView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        this._context = context;
        init();
    }

    public SketchHelpView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this._context = context;
        init();
    }

    private void init() {
        _paint = new Paint();
        _paint.setAntiAlias(true);
        _paint.setStrokeWidth(2);
        _paint.setStyle(Paint.Style.STROKE);
        _paint.setColor(Color.RED);
        _paint.setAlpha(_alpha);
        _paint.setTextSize(_textSize);
        _crossHair = BitmapFactory.decodeResource(_context.getResources(), R.mipmap.gmmfc_ic_cross_hair_v1);
        this.setWillNotDraw(false);
    }

    //橡皮绳的颜色
    int yellow = Color.parseColor("#ffff00");
    //分割面积颜色
    int blue = Color.parseColor("#62FFFF");
    String info;

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float cw = this.getWidth() / 2;
        float ch = this.getHeight() / 2;
        if (_text != null && _text.length > 0) {
            _paint.setStrokeWidth(2);
            _paint.setColor(Color.YELLOW);
            _paint.setTextSize( MeasureUtil.dp2px(getContext(),15));
            if(!"".equals(_text[0])){
                canvas.drawText(_text[0], cw - _crossHair.getWidth() / 2, ch - _crossHair.getHeight() * 2 / 3, _paint);
            }
            if(!"".equals(_text[1])){
                canvas.drawText(_text[1], cw - _crossHair.getWidth() / 2, ch - _crossHair.getHeight() * 4 / 3, _paint);
            }

        }
        if (_info != null && _info.size() > 0) {
            _paint.setStrokeWidth(1);
            _paint.setColor(Color.WHITE);
            _paint.setTextSize( MeasureUtil.dp2px(getContext(),10));
            Double area;
            for (PointF pointF : _info.keySet()) {
                area = _info.get(pointF);
                info = String.format("%.2f", area)+"亩";
                canvas.drawText(info, pointF.x - _paint.measureText(info) / 2, pointF.y, _paint);
            }
        }else {
            if (mLineInfo != null && mLineInfo.size() > 0) {
                _paint.setStyle(Paint.Style.FILL);
                _paint.setColor(Color.WHITE);
                _paint.setTextSize( MeasureUtil.dp2px(getContext(),15));
                _paint.setTextAlign(Paint.Align.CENTER);
                Double area;
                for (Path path : mLineInfo.keySet()) {
                    area = mLineInfo.get(path);
                    info = String.format("%.2f", area)+"m";
                    canvas.drawTextOnPath(info, path, 0,-8, _paint);
                }
            }
        }

        _paint.setStyle(Paint.Style.STROKE);
        _paint.setTextSize(_textSize);
        _paint.setColor(yellow);
        _paint.setStrokeWidth(3);
        if(_crossHair.isRecycled()){
            _crossHair = BitmapFactory.decodeResource(_context.getResources(), R.mipmap.gmmfc_ic_cross_hair_v1);
        }
        //中心点
        canvas.drawBitmap(_crossHair, cw - _crossHair.getWidth() / 2, ch - _crossHair.getHeight() / 2, _paint);
        //橡皮绳
        if (_isRelease) {
            if (_flag == DRAG_NORMAL) {
                dragHelpLine(canvas);
            }
        }
        //拖拽
        else {
            if (_flag == DRAG_POLYGON_POLYLINE) {
                dragPolygonDraw(canvas);
            }
            if (_flag == DRAW_TARGET) {
                drawTarget(canvas);
            }
            if (_flag == DRAG_POINT) {
                dragNodeDraw(canvas);
            }
        }
    }

    public PointF getScreenCenterPoint() {
        int w = this.getWidth();
        int h = this.getHeight();
        _screenCenterPoint = new PointF(w / 2, h / 2);
        return _screenCenterPoint;
    }

    /**
     * 橡皮绳
     *
     * @param points 1,2
     */
    public void setPoints(PointF... points) {
        _isRelease = true;
        _flag = DRAG_NORMAL;
        if (points == null) {
            _screenPoint1 = null;
            _screenPoint2 = null;
            return;
        }
        if (points.length == 1) {
            _screenPoint1 = points[0];
            _screenPoint2 = null;
        }
        if (points.length == 2) {
            _screenPoint1 = points[0];
            _screenPoint2 = points[1];
        }
        postInvalidate();
    }

    public void setText(String[] text) {
        this._text = text;
        postInvalidate();
    }


    private void dragHelpLine(Canvas canvas) {
        getScreenCenterPoint();
        if (_screenPoint1 == null) {
            _path = null;
        }
        /**
         * 橡皮线
         */
        if (_screenPoint1 != null && _screenPoint2 == null) {
            _path = new Path();
            _path.moveTo(_screenCenterPoint.x, _screenCenterPoint.y);
            _path.lineTo(_screenPoint1.x, _screenPoint1.y);
        }
        /**
         * 橡皮面
         */
        if (_screenPoint1 != null && _screenPoint2 != null) {
            _path = new Path();
            _path.moveTo(_screenPoint1.x, _screenPoint1.y);
            _path.lineTo(_screenCenterPoint.x, _screenCenterPoint.y);
            _path.lineTo(_screenPoint2.x, _screenPoint2.y);
        }
        if (_path != null) {
            canvas.drawPath(_path, _paint);
        }
    }


    /**
     * *********************************************已下非橡皮绳************************************************
     */
    /**
     * 初始化绘制参数
     */
    private void initPolygonDraw() {
        if (_oPointFS != null) {
            _mPointFS = new ArrayList<>();
            PointF _pf;
            for (int i = 0; i < _oPointFS.size(); i++) {
                _pf = _oPointFS.get(i);
                _mPointFS.add(new PointF(_pf.x + _vx, _pf.y + _vy));
            }
        }
    }

    /**
     * 绘制
     *
     * @param canvas
     */
    private void dragPolygonDraw(Canvas canvas) {
        initPolygonDraw();
        if (_mPointFS != null) {
            _path = new Path();
            for (int i = 0; i < _mPointFS.size(); i++) {
                if (i == 0) {
                    _path.moveTo(_mPointFS.get(i).x, _mPointFS.get(i).y);
                } else {
                    _path.lineTo(_mPointFS.get(i).x, _mPointFS.get(i).y);
                }
            }
            if (_isClose) {
                _path.close();
            }
            canvas.drawPath(_path, _paint);
        }
    }

    /**
     * 绘制拖拽多边形
     */
    public void startPolygonDraw() {
        _flag = DRAG_POLYGON_POLYLINE;
        _isRelease = false;
        postInvalidate();
    }

    public SketchHelpView setPointFS(List<PointF> pointFS) {
        this._oPointFS = pointFS;
        return this;
    }

    public SketchHelpView setClose(boolean close) {
        _isClose = close;
        return this;
    }

    public SketchHelpView setVT(float vx, float vy) {
        this._vx = vx;
        this._vy = vy;
        return this;
    }

    /**
     * *********************************************************************************************
     */
    public SketchHelpView setTargetPoint(PointF mbPoint) {
        this._mbPoint = mbPoint;
       // VibratorUtil.Vibrate(_context, 100);
        return this;
    }


    private void initTargetDraw() {
        _targetBitmap = BitmapFactory.decodeResource(_context.getResources(), R.mipmap.gmmfc_draw_tool_node_edit);
    }

    public void drawTargetView() {
        _flag = DRAW_TARGET;
        _isRelease = false;
        initTargetDraw();
        postInvalidate();
    }

    private void drawTarget(Canvas canvas) {
        int w = _targetBitmap.getWidth();
        int h = _targetBitmap.getHeight();
        _paint.setAlpha(_alpha);
        canvas.drawBitmap(_targetBitmap, _mbPoint.x - w / 2, _mbPoint.y - h, _paint);
    }


    public SketchHelpView dragNodeDraw(PointF aPoint, PointF bPoint, PointF cPoint, PointF mbPoint) {
        this._aPoint = aPoint;
        this._bPoint = bPoint;
        this._cPoint = cPoint;
        this._mbPoint = mbPoint;
        return this;
    }

    public void dragNodeDraw() {
        _flag = DRAG_POINT;
        postInvalidate();
    }

    // a*----------b*-----------c*
    private void dragNodeDraw(Canvas canvas) {
        drawTarget(canvas);
        _path = new Path();
        if (_aPoint != null && _bPoint != null && _cPoint != null) {
            _path.moveTo(_aPoint.x, _aPoint.y);
            _path.lineTo(_bPoint.x, _bPoint.y);
            _path.lineTo(_cPoint.x, _cPoint.y);
            _path.lineTo(_mbPoint.x, _mbPoint.y);
            _path.close();
        }
        if (_aPoint == null && _bPoint != null && _cPoint != null) {
            _path.moveTo(_bPoint.x, _bPoint.y);
            _path.lineTo(_mbPoint.x, _mbPoint.y);
        }
        if (_aPoint != null && _bPoint != null && _cPoint == null) {
            _path.moveTo(_bPoint.x, _bPoint.y);
            _path.lineTo(_mbPoint.x, _mbPoint.y);
        }
        canvas.drawPath(_path, _paint);
    }

    public PointF getEditPoint() {
        return _mbPoint;
    }

    /**
     * *********************************************************************************************
     */
    public int getFlag() {
        return _flag;
    }


    /**
     * *********************************************************************************************
     */
    public void clear(Boolean holdSplitArea) {
        _flag = DRAG_NOTHING;
        releaseDraw(holdSplitArea == null ? false : holdSplitArea);
        postInvalidate();
    }

    private void releaseDraw(boolean holdSplitArea) {
        _isRelease = true;
        if (_oPointFS != null) {
            _oPointFS.clear();
        }
        if (_mPointFS != null) {
            _mPointFS.clear();
        }
        if (_targetBitmap != null) {
            _targetBitmap.recycle();
        }
        if (!holdSplitArea) {
            _info = null;
        }
        mLineInfo = null;
        _text = null;
    }

    private Map<PointF, Double> _info;

    public void splitInfo(Map<PointF, Double> info) {
        this._info = info;
    }

    /**
     * 线段上填写的信息
     */
    private Map<Path,Double> mLineInfo;

    public void setLineInfo(Map<Path,Double> mLineInfo) {
        this.mLineInfo = mLineInfo;
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if(_targetBitmap!=null){
            _targetBitmap.recycle();
        }
        if(_crossHair!=null){
            _crossHair.recycle();
        }
    }
}

File: frame/map-frame-project/MapFrameCoreModuleV10/src/main/java/com/gradtech/mapframev10/core/sketch/SketchHelpView.kt
package com.gradtech.mapframev10.core.sketch

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.View
import com.gradtech.mapframev10.R
import com.mapbox.maps.ScreenCoordinate
import org.jetbrains.annotations.NotNull

/**
 * @ClassName SketchHelpView
 * @Description TODO
 * @Author: fs
 * @Date: 2022/8/5 9:33
 * @Version 2.0
 */
class SketchHelpView : View {

    private var mPaint: Paint = Paint();
    private var mCrossHair: Bitmap? = null;
    private var mDrawRubberLineDisabled: Boolean = true;

    private var mPeakPoint: PointF? = null;
    private var mOnePoint: PointF? = null;
    private var mTwoPoint: PointF? = null;
    private var mAreaMsg: String = "";
    private var mLengthMsg: String = "";
    private var mIsDrawAreaMsg = false;
    private var mIsDrawLengthMsg = false;

    constructor(context: Context?) : super(context) {
        init();
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init();
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init();
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init();
    }

    private fun init() {
        mPaint.setAntiAlias(true)
        mPaint.setStrokeWidth(2f)
        mPaint.setStyle(Paint.Style.STROKE)
        mPaint.setColor(Color.RED)
        mPaint.setAlpha(255)
        mPaint.setTextSize(30.0F)
        mCrossHair = BitmapFactory.decodeResource(context.resources, R.mipmap.gmmfc_ic_cross_hair_v1)

    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        Log.i("SketchHelpView", "" + width + "" + height)
        val cw = (this.width / 2).toFloat()
        val ch = (this.height / 2).toFloat()
        if (!mDrawRubberLineDisabled) {
            mPaint.color = Color.YELLOW
            val path = Path()
            path.moveTo(mOnePoint!!.x, mOnePoint!!.y)
            if (mOnePoint != null) {
                path.lineTo(mPeakPoint!!.x, mPeakPoint!!.y);
            }
            if (mTwoPoint != null) {
                path.lineTo(mTwoPoint!!.x, mTwoPoint!!.y);
            }
            canvas?.drawPath(path, mPaint)
        }
        //中心点
        canvas?.drawBitmap(mCrossHair, cw - mCrossHair!!.getWidth() / 2, ch - mCrossHair!!.getHeight() / 2, mPaint)
        mPaint.setColor(Color.YELLOW)
        canvas?.drawText(mLengthMsg, cw - mCrossHair!!.getWidth() / 2, ch - mCrossHair!!.getHeight()*2 / 3, mPaint)
        canvas?.drawText(mAreaMsg, cw - mCrossHair!!.getWidth() / 2, ch - mCrossHair!!.getHeight()*4 / 3, mPaint)
    }

    /**
     * 返回当前view的中心点
     */
    fun getScreenCenterPoint(): ScreenCoordinate {
        val w = this.width
        val h = this.height
        return ScreenCoordinate((w / 2).toDouble(), (h / 2).toDouble())
    }

    fun drawRubberLine(@NotNull onePoint: PointF, @NotNull twoPoint: PointF, peakPoint: PointF? = null) {
        val screenCenterPoint = getScreenCenterPoint()
        mPeakPoint = peakPoint
                ?: PointF(screenCenterPoint.x.toFloat(), screenCenterPoint.y.toFloat())
        mOnePoint = onePoint;
        mTwoPoint = twoPoint;
        Log.i("橡皮筋1", "" + mOnePoint!!.x + "," + mOnePoint!!.y);
        Log.i("橡皮筋2", "" + mTwoPoint!!.x + "," + mTwoPoint!!.y);
        if (mOnePoint!!.y < 0) {
            mOnePoint!!.y = 0f;
        }
        if (mOnePoint!!.x < 0) {
            mOnePoint!!.x = 0f;
        }
        if (mTwoPoint!!.y < 0) {
            mTwoPoint!!.y = 0f;
        }
        if (mTwoPoint!!.x < 0) {
            mTwoPoint!!.x = 0f;
        }
        mDrawRubberLineDisabled = false;
        this.postInvalidate()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        //mtargetBitmap.recycle()
        cleanup()
    }

    fun cleanup() {
        mCrossHair?.recycle()
    }

    /**
     * 设置面积
     */
    fun setAreaMsg(double2halfUp: Double) {
        if(mIsDrawAreaMsg){
            mAreaMsg = "面积:"+double2halfUp.toString()+"亩"
        }else{
            mAreaMsg = "";
        }
    }

    /**
     * 设置周长
     */
    fun setLengthMsg(double2halfUp: Double) {
        if(mIsDrawLengthMsg){
            mLengthMsg = "长度:"+double2halfUp.toString()+"米"
        }else{
            mLengthMsg = ""
        }
    }

    fun drawArea(isDrawAreaMsg: Boolean) {
        mIsDrawAreaMsg = isDrawAreaMsg
    }

    fun drawLength(isDrawLengthMsg: Boolean) {
        mIsDrawLengthMsg = isDrawLengthMsg
    }

}